src/ripple/server/impl/ServerImpl.cpp

```diff
@@ -33,75 +33,42 @@
 #include <time.h>
 
 namespace ripple {
-namespace HTTP {
 
 ServerImpl::ServerImpl (Handler& handler,
         boost::asio::io_service& io_service, beast::Journal journal)
-    : handler_ (handler)
-    , journal_ (journal)
-    , io_service_ (io_service)
-    , strand_ (io_service_)
-    , work_ (boost::in_place (std::ref(io_service)))
-    , hist_{}
+    : handler_(handler)
+    , j_(journal)
+    , io_service_(io_service)
+    , strand_(io_service_)
+    , work_(io_service_)
 {
 }
 
 ServerImpl::~ServerImpl()
 {
-    close();
-    {
-        // Block until all Door objects destroyed
-        std::unique_lock<std::mutex> lock(mutex_);
-        while (! list_.empty())
-            cond_.wait(lock);
-    }
+    // Handler::onStopped will not be called
+    work_ = boost::none;
+    ios_.close();
+    ios_.join();
 }
 
 void
 ServerImpl::ports (std::vector<Port> const& ports)
 {
     if (closed())
-        Throw<std::logic_error> ("ports() on closed HTTP::Server");
-    for(auto const& _ : ports)
-        if (! _.websockets())
-            std::make_shared<Door>(
-                io_service_, *this, _)->run();
-}
-
-void
-ServerImpl::onWrite (beast::PropertyStream::Map& map)
-{
-    std::lock_guard <std::mutex> lock (mutex_);
-    map ["active"] = list_.size();
+        Throw<std::logic_error> ("ports() on closed Server");
+    ports_.reserve(ports.size());
```

Comment:
Can `ports_` be non-empty? If so: `ports_.reserve(ports_.size() + ports.size())`. If not, then `assert(ports_.empty())`.